[NoGPU] Build MIOpen fails when GPU device does not exist

After enablement of #1024 MIOpen cannot build successfully on GPU-less platforms, such as WSL environment

[Task]
Differentiate "ROCMINFO FAILED, GPU TYPE UNKNOWN". and "ROCk module is NOT loaded, possibly no GPU devices".
Community can play with MIOpen build with WSL
MIOpen can be built on GPU-less platforms or servers

> @junliume
> Do we have any error logs on this issue? I don't quite understand how MIOpen should be build on WSL system.
> I assume that WSL does not contain ROCm and rocminfo. In this case PR#1081 will not help

It's mostly for dev convenience, previously worked but boke by GPU Detect PR.
You can technically install rocminfo on WSL, just the outcome would be:

❯ rocminfo
ROCk module is NOT loaded, possibly no GPU devices

without the "fix":

❯ make check -j16
... ...
-- Configuring done
-- Generating done
-- Build files have been written to: /mnt/c/Users/junliu14/MIOpen/build
*** FATAL: GPU DETECTION FAILED DURING CMAKE PHASE, CHECK CMAKE WARNINGS ***
make[3]: *** [test/CMakeFiles/gpu_detection_check.dir/build.make:58: test/CMakeFiles/gpu_detection_check] Error 1
make[2]: *** [CMakeFiles/Makefile2:9099: test/CMakeFiles/gpu_detection_check.dir/all] Error 2
make[2]: *** Waiting for unfinished jobs....
[  0%] Built target addkernels
[  0%] Built target sqlite_memvfs
make[1]: *** [CMakeFiles/Makefile2:11402: test/CMakeFiles/check.dir/rule] Error 2
make: *** [Makefile:5072: check] Error 2

--- a/test/CMakeLists.txt
+++ b/test/CMakeLists.txt
@@ -72,6 +72,7 @@ endif()
 # For HIP_NOGPU backend, GPU detection is not required and should be disabled.
 # Also we do not detect GPU when target GPU for testing is specified explicitly.
 set(MIOPEN_TEST_GPU_DETECTION_FAILED FALSE)
+set(MIOPEN_NO_GPU FALSE)
 if(NOT (MIOPEN_TEST_VEGA OR MIOPEN_TEST_GFX908 OR MIOPEN_TEST_GFX1030 OR MIOPEN_TEST_HIP_NOGPU))
     find_program(ROCMINFO
         NAMES rocminfo
@@ -88,7 +89,10 @@ if(NOT (MIOPEN_TEST_VEGA OR MIOPEN_TEST_GFX908 OR MIOPEN_TEST_GFX1030 OR MIOPEN_
             OUTPUT_VARIABLE ROCMINFO_OUTPUT
             RESULT_VARIABLE ROCMINFO_EXIT_STATUS
         )
-        if(NOT ROCMINFO_EXIT_STATUS EQUAL 0)
+        if(ROCMINFO_OUTPUT MATCHES "no GPU devices")
+            message(WARNING "ROCk module is NOT loaded, possibly no GPU devices")
+            set(MIOPEN_NO_GPU ON)         
+        elseif (NOT ROCMINFO_EXIT_STATUS EQUAL 0)
             message(WARNING "ROCMINFO FAILED, GPU TYPE UNKNOWN. Manually set respective MIOPEN_TEST_GFX* CMake variable to specify target GPU for testing.")
             set(MIOPEN_TEST_GPU_DETECTION_FAILED TRUE)
         elseif(ROCMINFO_OUTPUT MATCHES "gfx1030")
@@ -120,7 +124,7 @@ find_package(Threads REQUIRED)
 add_custom_target(check COMMAND ${CMAKE_CTEST_COMMAND} --output-on-failure -C ${CMAKE_CFG_INTDIR})
 add_custom_target(tests)
 
-if(MIOPEN_TEST_GPU_DETECTION_FAILED)
+if(MIOPEN_TEST_GPU_DETECTION_FAILED AND NOT (MIOPEN_NO_GPU))
     add_custom_target(gpu_detection_check COMMAND echo "*** FATAL: GPU DETECTION FAILED DURING CMAKE PHASE, CHECK CMAKE WARNINGS ***" COMMAND exit 1)
     add_dependencies(check gpu_detection_check)
 endif()
@@ -161,7 +165,7 @@ else()
 	if(MIOPEN_TEST_HALF)
 	    set(SKIP_ALL_EXCEPT_TESTS test_conv2d test_conv3d test_conv3d_extra test_immed_conv2d
                 test_immed_conv3d test_gru test_rnn_vanilla test_lstm test_gru_extra test_rnn_extra
-                test_lstm_extra )
+                test_lstm_extra)
 	elseif(MIOPEN_TEST_INT8)
 	    set(SKIP_ALL_EXCEPT_TESTS test_conv2d)
 	elseif(MIOPEN_TEST_BFLOAT16)
@@ -169,8 +173,14 @@ else()
     else()
         set(SKIP_ALL_EXCEPT_TESTS test_conv2d test_conv3d test_conv3d_extra test_immed_conv2d
                 test_immed_conv3d test_gru test_rnn_vanilla test_lstm test_gru_extra
-                test_rnn_extra test_lstm_extra )
+                test_rnn_extra test_lstm_extra)
 	endif()
+endif()
+
+if (MIOPEN_NO_GPU)
+    set(SKIP_ALL_EXCEPT_TESTS test_include_inliner test_kernel_build_params test_lstm test_lstm_dropout 
+            test_test_errors test_type_name test_tensor_test test_sqlite_perfdb test_sequences
+            test_pooling3d test_perfdb)
 endif()
 
 if(MIOPEN_TEST_GFX908)

